boltons/pathutils.py

```diff
@@ -0,0 +1,212 @@
+"""
+Functions for working with filesystem paths.
+
+The ``expandpath`` function expands the tilde to $HOME and environment variables
+to their values.
+
+The ``augpath`` function creates variants of an existing path without having to
+spend multiple lines of code spliting it up and stitching it back together.
+
+The ``shrinkuser`` function replaces your home directory with a tilde.
+
+The ``userhome`` function reports the home directory of the current user of the
+operating system.
+"""
+from __future__ import print_function
+
+from os.path import dirname
+from os.path import exists
+from os.path import expanduser
+from os.path import expandvars
+from os.path import join
+from os.path import normpath
+from os.path import split
+from os.path import splitext
+import os
+import sys
+
+
+__all__ = [
+    'augpath', 'shrinkuser', 'userhome', 'expandpath',
+]
+
+
+def augpath(path, suffix='', prefix='', ext=None, base=None, dpath=None,
+            multidot=False):
+    """
+    Create a new path with a different extension, basename, directory, prefix,
+    and/or suffix.
+
+    A prefix is inserted before the basename. A suffix is inserted
+    between the basename and the extension. The basename and extension can be
+    replaced with a new one. Essentially a path is broken down into components
+    (dpath, base, ext), and then recombined as (dpath, prefix, base, suffix,
+    ext) after replacing any specified component.
+
+    Args:
+        path (PathLike): a path to augment
+        suffix (str, default=''): placed between the basename and extension
+        prefix (str, default=''): placed in front of the basename
+        ext (str, optional): if specified, replaces the extension
+        base (str, optional): if specified, replaces the basename without extension
+        dpath (PathLike, optional): if specified, replaces the directory
+        multidot (bool, default=False): if False, everything after the last dot
+            in the basename is the extension. If True, everything after the first
+            dot in the basename is the extension.
+
+    Returns:
+        PathLike: augmented path
+
+    Example:
+        >>> path = 'foo.bar'
+        >>> suffix = '_suff'
+        >>> prefix = 'pref_'
+        >>> ext = '.baz'
+        >>> newpath = augpath(path, suffix, prefix, ext=ext, base='bar')
+        >>> print('newpath = %s' % (newpath,))
+        newpath = pref_bar_suff.baz
+
+    Example:
+        >>> augpath('foo.bar')
+        'foo.bar'
+        >>> augpath('foo.bar', ext='.BAZ')
+        'foo.BAZ'
+        >>> augpath('foo.bar', suffix='_')
+        'foo_.bar'
+        >>> augpath('foo.bar', prefix='_')
+        '_foo.bar'
+        >>> augpath('foo.bar', base='baz')
+        'baz.bar'
+        >>> augpath('foo.tar.gz', ext='.zip', multidot=True)
+        foo.zip
+        >>> augpath('foo.tar.gz', ext='.zip', multidot=False)
+        foo.tar.zip
+        >>> augpath('foo.tar.gz', suffix='_new', multidot=True)
+        foo_new.tar.gz
```

Comment:
Ah, thanks for this example. I was wondering what `multidot` meant practically, and I see it now.

Reply:
Do you have any ideas for a better name than `multidot`? I was thinking `dots` as a shorthand, but I'm not sure if it becomes unclear as to what its doing.

Or perhaps `ndots`, where you specify the maximum number of dots allowed in an extension?

Reply:
hmm, been thinking about this one. I can't remember whether I knew the number of dots back when I was splitting extensions last. I could see `ndots` or `maxdots` though. 

If it stays a bool flag, then the only names that pop into my head are `greedy` and `eager`. Multidot isn't that bad though.

Reply:
Thanks for giving this some thought. I guess it make sense to leave it as a boolean with the current name, given that nothing comes to mind easily. 

The main issue I have with the name is that it is 8 characters plus the 5 characters for `=True`. I would prefer to have something slightly more compact because I often find that setting this flag myself going over the 80 character line limit when I use this function, but that is a first world problem. 

I think the other part that bugs me about this is that in general it's not possible to know whether text after the first dot is meant to be a file extension or is actually part of the file name. Sure there are heuristics (e.g. look for `.tar.gz`), but these break down easily (e.g. `.json.custom.backup`). I also agree that I typically don't know the exact number of dots that are possible I'm expecting when processing a filepath, but I typically will know if dots are allowed in the filename or not, so it might be better to opt for the simpler boolean API.